File: lua/cn_2en_pinyin.lua
local stringUtil = require('./lib/string_util')

local dict = {}

local function isChineseWord(word)
    -- UTF-8: 3 bytes per Chinese character.
    return word:match("[\228-\233][\128-\191][\128-\191]")
end

local Filter = {}

-- line format: 不可数集 [bù kě shuò jí] /uncountable set (math.)/
function Filter.parseLine(line)
    local word, pinyin, desc = line:match("^(.-) %[(.-)%] /(.+)/.*$")
    if not word or word:len() < 1 then return end

    return {
        word = word,
        info = pinyin .. '|' .. desc
    }
end

local splitter = "\\\\"

local isFilterInitialized = false -- 重启输入法有时会初始化插件两次，导致字典被加载两次。绕过之。
function Filter.init(env)
    if isFilterInitialized then
        -- log.writeLog('isFilterInitialized = true')
        return
    end

    local fileAbsolutePath = rime_api.get_user_data_dir() .. "/lua/data/cedict_fixed.u8"
    local file = io.open(fileAbsolutePath, "r")
    if file == nil then return end

    for line in file:lines() do
        local w = Filter.parseLine(line)
        -- there might be multiple entries of the same word, store them together
        if w then
            local previous = dict[w.word]
            if previous then
                dict[w.word] = previous .. splitter .. w.info  -- Append if already exists
            else
                dict[w.word] = w.info
            end
        end
    end
    file:close()
    isFilterInitialized = true
end

local function tryTranslateToEnglish(typedText, engine, candidateList)
    if #typedText < 3 then return end

    local hintCodes = 'ⁿᵃᵇᶜᵈᵉᶠᵍʰⁱʲᵏˡᵐᵒᵖʳˢᵗᵘᵛʷˣʸᶻ'
    local hintKeys = 'nabcdefghijklmoprstuvwxyz'
    if typedText:find('/e') then
        -- '/e' 启用汉译英上屏，注释上提示汉译英候选项的辅助码 => '/ea', '/eb', ...
        local idxHint = 1
        for i = 1, #candidateList do
            local item = candidateList[i]
            if item.en and idxHint <= #hintKeys then
                local hint = stringUtil.utf8_sub(hintCodes, idxHint, idxHint)
                item.candidate.comment = '⇖' .. hint .. item.candidate.comment
                idxHint = idxHint + 1
            end
        end
    end

    if typedText:sub(#typedText - 2, #typedText - 1) ~= '/e' then return end

    local lastChar = typedText:sub(#typedText)
    local idx = hintKeys:find(lastChar) or 100 --'/en' => 1st, '/ea' => 2nd, '/eb' => 3rd, ...
    local pickingItem = candidateList[idx]
    if not pickingItem then return end

    local arr = stringUtil.split(pickingItem.en, '/')
    if #arr == 1 then -- 唯一翻译，直接上屏
        engine:commit_text(arr[1])
        engine.context:clear()
    else -- 多种翻译，插入候选
        for i = 1, #arr, 1 do
            -- 量词部分不要候选: 上衣 [shàng yī] /jacket/upper outer garment/CL:件[jiàn]/
            if not arr[i]:find('CL') then
                yield(Candidate("en", 0, #typedText, arr[i], "翻译自：" .. pickingItem.candidate.text))
            end
        end
        yield(pickingItem.candidate) -- 把对应的汉语候选项移到最上面
    end
end

local function tryTranslateToPinyin(typedText, engine, candidateList)
    if #typedText < 3 then return end

    local hintCodes = 'ʸᵃᵇᶜᵈᵉᶠᵍʰⁱʲᵏˡᵐⁿᵒᵖʳˢᵗᵘᵛʷˣᶻ'
    local hintKeys = 'yabcdefghijklmnoprstuvwxz'
    if typedText:find('/p') then
        -- '/p' 启用拼音上屏，各候选项的注释上提示辅助码 => '/py', '/pa', ...
        local idxHint = 1
        for i = 1, #candidateList do
            local item = candidateList[i]
            if item.en and idxHint <= #hintKeys then
                local hint = stringUtil.utf8_sub(hintCodes, idxHint, idxHint)
                item.candidate.comment = '⇖' .. hint .. item.candidate.comment
                idxHint = idxHint + 1
            end
        end
    end

    if typedText:sub(#typedText - 2, #typedText - 1) ~= '/p' then return end

    local lastChar = typedText:sub(#typedText)
    local idx = hintKeys:find(lastChar) or 100 --'/py' => 1st, '/pa' => 2nd, '/pb' => 3rd, ...
    local pickingItem = candidateList[idx]
    if not pickingItem then return end

    local arr = stringUtil.split(pickingItem.pinyin, ' ')
    if #arr == 1 then -- 单字拼音，直接上屏
        engine:commit_text(arr[1])
        engine.context:clear()
    else
        -- 词组拼音插入候选
        yield(Candidate("py", 0, #typedText, pickingItem.pinyin, "拼音：" .. pickingItem.candidate.text))

        for i = 1, #arr, 1 do -- 拆开成单字后插入候选
            yield(Candidate("py", 0, #typedText, arr[i], "拼音：" .. pickingItem.candidate.text))
        end

        yield(pickingItem.candidate) -- 把对应的汉语候选项移到最上面
    end
end

-- 根据字典拼音匹配的权重，重排对应的候选项
-- candidatesToYield 里没有查找到拼音的候选项，原位不动
local function sortCandidatesByPinyin(candidatesToYield, candidatesWithCedict)
    -- 先记录查找到拼音的候选项在 candidatesToYield 里的位置
    local idxArray = {}
    for i = 1, #candidatesWithCedict do
        idxArray[i] = candidatesWithCedict[i].idx
    end
    -- 根据权重重排 candidatesWithCedict
    table.sort(candidatesWithCedict, function(a, b)
        return a.weight > b.weight
    end)
    -- 根据 candidatesWithCedict 的顺序，就地重排 candidatesToYield
    for i = 1, #candidatesWithCedict do
        candidatesToYield[idxArray[i]] = candidatesWithCedict[i].candidate
    end
end

function Filter.func(input, env)
    local typedText = env.engine.context.input
    -- 忽略 / 后的字符，如 `zhangk/e` 只用 `zhangk` 来匹配
    -- 以免选择汉译英时，按下 `/e` 又把 `展开` 顶到最上面
    local prefix = typedText:gsub('(/.*)', '')

    local candidatesToYield = {}
    local candidatesWithCedict = {}
    local newCandidates = {}  -- 多音字的第一项放在 candidatesToYield 里，其余放在 newCandidates 里

    local idx = 0
    for cand in input:iter() do
        idx = idx + 1
        if isChineseWord(cand.text) then
            local info = dict[cand.text]
            if info then
                local arr = stringUtil.split(info, splitter)
                for i = 1, #arr, 1 do
                    local pinyin, en = arr[i]:match("^(.-)|(.+)$")
                    -- 仅在输入字符和拼音前序匹配时，显示拼音和英文注释，避免多音字和模糊音候选项刷屏
                    local unaccentedPinyin = stringUtil.unaccent(pinyin:gsub(' ', ''))
                    local isInputMatchingPinyin = unaccentedPinyin:sub(1, #prefix) == prefix
                    if isInputMatchingPinyin then
                        local comment = '〖' .. pinyin .. '〗' .. en
                        if i == 1 then
                            cand.comment = comment
                            table.insert(candidatesWithCedict, {
                                candidate = cand,
                                pinyin = pinyin,
                                en = en,
                                weight = unaccentedPinyin == prefix and 10000 - idx or 1000 - idx,
                                idx = idx,
                            })
                        else
                            local candidate = Candidate("cn", 0, #prefix, cand.text, comment)
                            newCandidates[cand.text] = newCandidates[cand.text] or {}
                            table.insert(newCandidates[cand.text], candidate)
                        end
                    end
                end
            end
        end
        table.insert(candidatesToYield, cand)
    end

    tryTranslateToEnglish(typedText, env.engine, candidatesWithCedict)
    tryTranslateToPinyin(typedText, env.engine, candidatesWithCedict)

    sortCandidatesByPinyin(candidatesToYield, candidatesWithCedict)

    for _, v in ipairs(candidatesToYield) do
        yield(v)
        local newCandidatesWithTheSameWord = newCandidates[v.text] or {}
        for _, candidate in ipairs(newCandidatesWithTheSameWord) do
            yield(candidate)
        end
    end
end

return Filter

File: js/cn2en_pinyin.js
/**
 * 中文输入法插件，提供以下功能：
 * 1. 为中文候选项添加拼音注解和英文释义
 * 2. 支持通过 /p 快捷键选择拼音（如 /py 选择第一个候选项的拼音，/pa 选择第二个候选项的拼音）
 * 3. 支持通过 /e 快捷键选择英文翻译（如 /en 选择第一个候选项的翻译，/ea 选择第二个候选项的翻译）
 *
 * -------------------------------------------------------
 * 使用 JavaScript 实现，适配 librime-qjs 插件系统。
 * @author https://github.com/HuangJian
 */

import { isChineseWord } from './lib/string.js'

/**
 * 设置查找拼音和英文释义的候选项数量上限
 * @type {number}
 * @description 为了避免性能问题，只查找前面指定数量的候选项
 */
const sizeToLookupEnglish = 200 // 只查找前面 200 项的拼音和英文释义

/**
 * 英文翻译选择的上标提示符号
 * @type {string}
 * @description 用于在候选项注释中显示的上标字符，提示用户可通过快捷键选择对应的英文翻译
 */
const enHintCodes = 'ⁿᵃᵇᶜᵈᵉᶠᵍʰⁱʲᵏˡᵐᵒᵖʳˢᵗᵘᵛʷˣʸᶻ'

/**
 * 英文翻译选择的快捷键
 * @type {string}
 * @description 与 enHintCodes 一一对应的快捷键，用于选择对应的英文翻译
 */
const enHintKeys = 'nabcdefghijklmoprstuvwxyz'

/**
 * 拼音选择的上标提示符号
 * @type {string}
 * @description 用于在候选项注释中显示的上标字符，提示用户可通过快捷键选择对应的拼音
 */
const pyHintCodes = 'ʸᵃᵇᶜᵈᵉᶠᵍʰⁱʲᵏˡᵐⁿᵒᵖʳˢᵗᵘᵛʷˣᶻ'

/**
 * 拼音选择的快捷键
 * @type {string}
 * @description 与 pyHintCodes 一一对应的快捷键，用于选择对应的拼音
 */
const pyHintKeys = 'yabcdefghijklmnoprstuvwxz'

/**
 * 汉译英过滤器
 * @implements {FastFilter}
 */
export class Cn2EnFilter {
  /**
   * LevelDb对象，用于存储和查询中文词汇的拼音和英文释义
   * @type {LevelDb}
   */
  levelDb = null

  /**
   * 初始化插件
   * @param {Environment} env - 环境对象，包含用户数据目录、文件操作等功能
   * @description 加载字典数据，优先从二进制文件加载以提高性能，如果二进制文件不存在则从文本文件加载并生成二进制文件
   */
  constructor(env) {
    console.log('cn2en_pinyin filter init')

    // @ts-expect-error
    this.levelDb = env.levelDb || new LevelDb()

    // @ts-expect-error for unit test
    const txtPath = env.cn2enTextFilePath || `${env.userDataDir}/js/data/cedict_fixed.u8`

    // @ts-expect-error for unit test
    const binPath = env.en2cnBinaryFilePath || `${env.userDataDir}/js/data/cn2en.ldb`

    let tick = Date.now()
    if (env.fileExists(binPath)) {
      this.levelDb.loadBinaryFile(binPath)
      console.log(`cn2en_pinyin filter: load dict from binary file takes: ${Date.now() - tick}ms`)
    } else {
      this.levelDb.loadTextFile(txtPath, {
        lines: 119000,
        charsToRemove: '\r',
        onDuplicatedKey: 'Overwrite',
      })
      this.levelDb.saveToBinaryFile(binPath) // leveldb 必须要保存成二进制文件，否则无法使用查找功能
      console.log(`cn2en_pinyin filter: saved dict to a binary file takes: ${Date.now() - tick}ms`)
    }
  }

  /**
   * 插件终止函数
   * @description 在输入法退出或重新部署时调用，用于清理资源
   */
  finalizer() {
    console.log('cn2en_pinyin filter finit')
    this.levelDb?.close()
  }

  /**
   * Check if the filter is applicable in the current context
   * @param {Environment} env - The Rime environment
   * @returns {boolean} True if the filter is applicable, otherwise false
   */
  isApplicable(env) {
    return env.engine.context.input.length > 1
  }

  /**
   * 候选项过滤器主函数
   * @param {CandidateIterator} iter - 候选项迭代器，用于遍历候选项
   * @param {Environment} env - 环境对象，包含引擎上下文等信息
   * @returns {Generator<Candidate, CandidateIterator | void>} 处理后的候选项
   * @description 为中文候选项添加拼音注解和英文释义，并处理拼音和英文翻译的快捷选择功能
   */
  *filter(iter, env) {
    const input = env.engine.context.input

    const processed = []
    // 只查找前面 sizeToLookupPinyin 个候选词，提高性能
    for (let idx = 0, candidate; idx < sizeToLookupEnglish && (candidate = iter.next()); idx++) {
      if (
        candidate.text.length > 10 || // 超过10个字的词不查找
        !isChineseWord(candidate.text) || // not a Chinese word
        candidate.comment.includes('〖') || // 已经查过了
        false
      ) {
        processed.push(candidate)
        continue
      }

      const info = this.levelDb.find(candidate.text)
      if (info) {
        extractCandidatesByInfo(candidate, info, input).forEach((it) => {
          processed.push(it)
        })
      } else {
        processed.push(candidate)
      }
    }

    hintToPickEnglish(processed, input)
    tryPrependOrCommitEnglish(processed, input, env.engine)

    hintToPickPinyin(processed, input)
    tryPrependOrCommitPinyin(processed, input, env.engine)

    yield* processed
    return iter
  }
}

/**
 * 根据词典信息提取候选项
 * @param {Object} candidate - 原始候选项对象
 * @param {string} info - 从字典中查询到的信息，包含拼音和英文释义
 * @param {string} inputCode - 用户输入的编码
 * @returns {Array} 处理后的候选项数组
 * @description 解析字典中的拼音和英文释义信息，为候选项添加注解。如果一个词有多个拼音或释义，会创建多个候选项
 */
function extractCandidatesByInfo(candidate, info, inputCode) {
  // 去掉白霜拼音方案设置在注解中的拼音，但保留其它插件的注解，如 pin_cand_filter 的 📌，is_in_user_dict 的 ∞/*
  const prevComment = candidate.comment?.replace(/(［.+)］/g, '')

  // format: [diǎn diǎn]Diandian (Chinese microblogging and social networking website)||[diǎn diǎn]point/speck
  const ret = info
    .split('||')
    .map((item, idx) => {
      const [_, pinyin, en] = /^\[(.*?)\](.*)$/.exec(item) || []
      if (!pinyin) return null

      const comment = prevComment + '〖' + pinyin + '〗' + en
      let theCandidate = candidate
      if (idx > 0) {
        theCandidate = new Candidate('cn', 0, inputCode.length, candidate.text, comment)
      } else {
        theCandidate.comment = comment
      }
      theCandidate.pinyin = pinyin
      theCandidate.en = en

      return theCandidate
    })
    .filter((it) => it)
  return ret.length > 0 ? ret : [candidate]
}

/**
 * 为候选项添加英文翻译选择的提示标记
 * @param {Array} candidates - 候选项数组
 * @param {string} input - 用户输入的原始字符串
 * @description 当用户输入以 /e 结尾时，为含有英文翻译的候选项添加上标提示符号，便于用户通过快捷键选择
 */
function hintToPickEnglish(candidates, input) {
  if (input.length < 3 || !/\/e.?$/.test(input)) return

  // '/e' 启用汉译英上屏，注释上提示汉译英候选项的辅助码 => '/en', '/ea', '/eb', ...
  let idxHint = 0
  for (let i = 0; i < candidates.length; i++) {
    const item = candidates[i]
    if (item.en && idxHint < enHintKeys.length) {
      item.comment = '⇖' + enHintCodes[idxHint] + item.comment
      if (idxHint++ >= enHintKeys.length) break
    }
  }
}

/**
 * 处理英文翻译的选择和上屏
 * @param {Array} candidates - 候选项数组
 * @param {string} input - 用户输入的原始字符串
 * @param {Object} engine - 输入法引擎对象
 * @description 当用户输入以 /en、/ea 等快捷键结尾时，选择对应的英文翻译并上屏或添加到候选项列表
 */
function tryPrependOrCommitEnglish(candidates, input, engine) {
  const inputSize = input.length
  if (input.substring(inputSize - 3, inputSize - 1) !== '/e') return

  const lastChar = input.substring(inputSize - 1, inputSize)
  const idx = enHintKeys.indexOf(lastChar) // '/en' => 1st, '/ea' => 2nd, '/eb' => 3rd, ...
  if (idx < 0) return

  const hintCode = '⇖' + enHintCodes[idx]
  const pickingItem = candidates.find((it) => it.comment?.startsWith(hintCode))
  if (!pickingItem) return

  const rightBracketPos = pickingItem.comment.indexOf('〗')
  const arr = pickingItem.comment
    .substring(rightBracketPos + 1)
    .split('/')
    .filter((it) => !it.includes('CL')) // CL 表示量词，不候选

  if (arr.length === 1) {
    // 唯一翻译，直接上屏
    engine.commitText(arr[0])
    engine.context.clear()
  } else {
    // 多种翻译，插入候选
    candidates.unshift(pickingItem) // 把对应的汉语候选项移到最上面
    const text = pickingItem.text
    arr.reverse().forEach((it) => {
      candidates.unshift(new Candidate('en', 0, inputSize + 2, it, '翻译自：' + text))
    })
  }
}

/**
 * 为候选项添加拼音选择的提示标记
 * @param {Array} candidates - 候选项数组
 * @param {string} input - 用户输入的原始字符串
 * @description 当用户输入以 /p 结尾时，为含有拼音的候选项添加上标提示符号，便于用户通过快捷键选择
 */
function hintToPickPinyin(candidates, input) {
  if (input.length < 3 || !/\/p.?$/.test(input)) return

  // '/p' enables pinyin selection with hint codes => '/py', '/pa', '/pb', ...
  let idxHint = 0
  for (let i = 0; i < candidates.length; i++) {
    const item = candidates[i]
    if (item.pinyin && idxHint < pyHintKeys.length) {
      item.comment = '⇖' + pyHintCodes[idxHint] + item.comment
      if (idxHint++ >= pyHintKeys.length) break
    }
  }
}

/**
 * 处理拼音的选择和上屏
 * @param {Array} candidates - 候选项数组
 * @param {string} input - 用户输入的原始字符串
 * @param {Object} engine - 输入法引擎对象
 * @description 当用户输入以 /py、/pa 等快捷键结尾时，选择对应的拼音并上屏或添加到候选项列表
 */
function tryPrependOrCommitPinyin(candidates, input, engine) {
  const inputSize = input.length
  if (input.substring(inputSize - 3, inputSize - 1) !== '/p') return

  const lastChar = input.substring(inputSize - 1, inputSize)
  const idx = pyHintKeys.indexOf(lastChar) // '/py' => 1st, '/pa' => 2nd, '/pb' => 3rd, ...
  if (idx < 0) return

  const hintCode = '⇖' + pyHintCodes[idx]
  const pickingItem = candidates.find((it) => it.comment?.startsWith(hintCode))
  if (!pickingItem) return

  const arr = pickingItem.pinyin.split(' ')
  if (arr.length === 1) {
    // Single pinyin, commit directly
    engine.commitText(arr[0])
    engine.context.clear()
  } else {
    // Multiple pinyin syllables, insert as candidates
    candidates.unshift(pickingItem) // Move the corresponding Chinese candidate to the top
    arr.reverse().forEach((it) => {
      candidates.unshift(new Candidate('py', 0, inputSize + 2, it, '拼音：' + pickingItem.text))
    })
    candidates.unshift(new Candidate('py', 0, inputSize + 2, pickingItem.pinyin, '拼音：' + pickingItem.text))
  }
}
